src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,260 @@
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=cache_hits,
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=cache_misses,
+        )
+        mapped_results.update(cache_keys.get_mapped_results())
+
+        db_read_keys = cache_keys.get_unmapped_keys()
+
+        if not db_read_keys.as_strings():
+            return mapped_results
+
+        # Step 2: Lookup ids in the db and add them to the
+        # db_read KeyCollection results.
+        for db_obj in self._get_db_records(db_read_keys):
+            db_read_keys.add_key_result((db_obj.organization_id, db_obj.string), db_obj.id)
+
+        db_read_key_results = db_read_keys.get_mapped_results()
+
+        if len(db_read_key_results.keys()) > 0:
+            results_to_cache = db_read_keys.get_mapped_key_strings_to_ints()
+            indexer_cache.set_many(results_to_cache)
+            mapped_results.update(db_read_key_results)
+
+        db_write_keys = db_read_keys.get_unmapped_keys()
+
+        if not db_write_keys.as_strings():
+            return mapped_results
+
+        # Step 3: Create new records in the db and add them to the
+        # db_write KeyCollection results.
+        new_records = []
+        for write_pair in db_write_keys.as_tuples():
+            organization_id, string = write_pair
+            new_records.append(
+                StringIndexerTable(organization_id=int(organization_id), string=string)
+            )
+
+        with metrics.timer("sentry_metrics.indexer.pg_bulk_create"):
+            # We use `ignore_conflicts=True` here to avoid race conditions where metric indexer
+            # records might have be created between when we queried in `bulk_record` and the
+            # attempt to create the rows down below.
+            StringIndexerTable.objects.bulk_create(new_records, ignore_conflicts=True)
+
+        for db_obj in self._get_db_records(db_write_keys):
+            db_write_keys.add_key_result((db_obj.organization_id, db_obj.string), db_obj.id)
+
+        db_write_key_results = db_write_keys.get_mapped_results()
+
+        if len(db_write_key_results) > 0:
+            results_to_cache = db_write_keys.get_mapped_key_strings_to_ints()
+            indexer_cache.set_many(results_to_cache)
+            mapped_results.update(db_write_key_results)
+
+        return mapped_results
+
+    def record(self, org_id: int, string: str) -> int:
+        """Store a string and return the integer ID generated for it"""
+        result = self.bulk_record({org_id: {string}})
+        return result[org_id][string]
+
+    def resolve(self, org_id: int, string: str) -> Optional[int]:
+        """Lookup the integer ID for a string.
+
+        Returns None if the entry cannot be found.
+
+        """
+        key = f"{org_id}:{string}"
+        result = indexer_cache.get(key)
```

Comment:
Please use the key collection even if you have only one entry. See above on why we should not scatter packing and unpacking in the code.